src/cache_cf.cc

```diff
@@ -298,139 +299,144 @@ skip_ws(const char* s)
+    static const auto kidIdentifier = ToSBuf(KidIdentifier);
+    SubstituteMacro(buf, macroServiceName, service_name);
+    SubstituteMacro(buf, macroProcessName, TheKidName);
+    SubstituteMacro(buf, macroProcessNumber, kidIdentifier);
 }
 
-static void
-ProcessMacros(char*& line, int& len)
+/// extracts the next optional token surrounded by optional space
+static std::optional<SBuf>
+ExtractOptionalToken(Parser::Tokenizer &tk)
 {
-    SubstituteMacro(line, len, "${service_name}", service_name.c_str());
-    SubstituteMacro(line, len, "${process_name}", TheKidName.c_str());
-    SubstituteMacro(line, len, "${process_number}", xitoa(KidIdentifier));
+    const auto &spaceChars = CharacterSet::libcSpace();
+    (void)tk.skipAll(spaceChars);
+
+    SBuf token;
+    static const auto tokenChars = spaceChars.complement("ExtractOptionalToken() token chars");
+    if (!tk.prefix(token, tokenChars))
+        return std::nullopt;
+
+    (void)tk.skipAll(spaceChars);
+    return token;
 }
 
-static void
-trim_trailing_ws(char* str)
+/// extracts the next required token surrounded by optional space
+static SBuf
+ExtractToken(const char * const description, Parser::Tokenizer &tk)
 {
-    assert(str != nullptr);
-    unsigned i = strlen(str);
-    while ((i > 0) && xisspace(str[i - 1]))
-        --i;
-    str[i] = '\0';
+    if (const auto token = ExtractOptionalToken(tk))
+        return *token;
+    throw TextException(ToSBuf("cannot find ", description), Here());
 }
 
-static const char*
-FindStatement(const char* line, const char* statement)
+/// extracts the next token if it matches the given keyword
+/// \returns whether the token was extracted
+static bool
+SkipKeyword(const SBuf &keyword, Parser::Tokenizer &tk)
 {
-    assert(line != nullptr);
-    assert(statement != nullptr);
-
-    const char* str = skip_ws(line);
-    unsigned len = strlen(statement);
-    if (strncmp(str, statement, len) == 0) {
-        str += len;
-        if (*str == '\0')
-            return str;
-        else if (xisspace(*str))
-            return skip_ws(str);
+    const auto savedTk = tk;
+    if (const auto token = ExtractOptionalToken(tk)) {
+        if (token == keyword)
+            return true;
     }
-
-    return nullptr;
+    tk = savedTk;
+    return false;
 }
 
-static bool
-StrToInt(const char* str, long& number)
+/// extracts the next input token (skipping any leading space) and interprets it as
+/// a signed integer (in decimal, hex, or octal base per Parser::Tokenizer::int64())
+static int64_t
+ExtractNumber(const char * const description, Parser::Tokenizer &tk)
 {
-    assert(str != nullptr);
-
-    char* end;
-    number = strtol(str, &end, 0);
-
-    return (end != str) && (*end == '\0'); // returns true if string contains nothing except number
+    auto numberParser = Parser::Tokenizer(ExtractToken(description, tk));
+    int64_t result = 0;
+    if (!numberParser.int64(result, 0, true))
```

Comment:
This default '0' (guess) does not work in my configuration `if ${process_number} = 1`:
```
FATAL: bad configuration: cannot parse left-hand operand of an equality condition
    exception location: cache_cf.cc(413) ExtractNumber
```
Parser::Tokenizer() gets zero '0' token but int64() treats it as octal notation.

Reply:
> Parser::Tokenizer() gets zero '0' token but int64() treats it as octal notation.

Indeed. Glad you found this Tokenizer bug! I have posted a fix at #1842. I am marking this PR as blocked on that PR.

Reply:
> Glad you found this Tokenizer bug! I have posted a fix at #1842. I am marking this PR as blocked on that PR.

I have merged that fix and removed S-waiting-for-PR label.